ml4gw/transforms/decimator.py

```diff
@@ -0,0 +1,195 @@
+import torch
+
+
+class Decimator(torch.nn.Module):
+    """
+    Downsample/decimate timeseries according to a schedule.
+
+    The schedule is a tensor of shape (N, 3), where each row is:
+    [start_time, end_time, target_sample_rate]
+
+    Example default schedule (seconds, Hz):
+    [[0, 40, 256],   # first 40s at 256 Hz
+     [40, 58, 512],  # next 18s at 512 Hz
+     [58, 60, 2048]] # last 2s at full 2048 Hz
+
+    Strain: Tensor of shape (B, C, T), where
+    B=batch size, C=channels, T=time.
+
+    Functions:
+        build_variable_indices(sample_rate, schedule, device)
+            Compute the sample indices to keep from the full strain.
+        split_by_schedule(strain, schedule)
+            Split a strain into segments according to the schedule.
+    """
+
+    def __init__(
+        self,
+        sample_rate: int = None,
+        schedule: torch.Tensor = None,
+        device=None,
+    ):
+        super().__init__()
+        self.sample_rate = sample_rate
+        self.schedule = schedule
+        self.device = device
+
+        if sample_rate is not None and schedule is not None:
+            self._validate_inputs(sample_rate, schedule)
+            self.idx = self.build_variable_indices(
+                sample_rate, schedule, device
+            )
+        else:
+            self.idx = None
+
+    def _validate_inputs(
+        self,
+        sample_rate: int,
+        schedule: torch.Tensor,
+        strain: torch.Tensor = None,
+    ):
+        """
+        Validate inputs for compatibility.
+        If `strain` is None, only validates schedule and sample_rate.
+        """
+        if not isinstance(sample_rate, int) or sample_rate <= 0:
```

Comment:
Testing for type and for positivity is probably overkill. There's nothing wrong with doing so, but it's reasonable to expect the user to do something sensible here. Also, in your testing, you should make sure that these errors actually get raised. You can see the process for doing so in some of our other tests, and you can see the testing coverage report in the comments on this PR; let me know if you have any questions about it.